### PR TITLE
test(strategies): implement tests

### PR DESCRIPTION
**Pull request by BioPhoton**
This PR implements a test set for all but scheduleCd :)

### CODE REVIEW

File: libs/template/spec/render-strategies/render-strategies/noop.strategy.spec.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import { from, of } from 'rxjs';
+import createSpy = jasmine.createSpy;
+
+const noopStrategyName = 'noop';
+
+describe('noop Strategy', () => {
+  beforeAll(() => mockConsole());
+
+  it('should be present in strategies map', () => {
+    const strategy = getStrategies(getMockNoopStrategyConfig())[noopStrategyName];
```
Comment (Karnaukhov-kh) [inline libs/template/spec/render-strategies/render-strategies/noop.strategy.spec.ts:14]: Can this be a file-level const and assignment placed in `beforeEach`? (applicable for all strategy specs files)

File: libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
+// tslint:disable-next-line:nx-enforce-module-boundaries
+import { mockConsole } from '@test-helpers';
+import { getStrategies } from '@rx-angular/template';
+import { getMockNativeStrategyConfig } from '../../fixtures';
+import { from, of } from 'rxjs';
+import createSpy = jasmine.createSpy;
+
+const nativeStrategyName = 'native';
```
Comment (Karnaukhov-kh) [inline libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts:8]: If we go with a generic approach, maybe just `strategyName` / `testedStrategyName` ?

File: libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      complete: () => {
+        expect(cfg.cdRef.markForCheck).toHaveBeenCalledTimes(6);
+        expect(cfg.cdRef.detectChanges).toHaveBeenCalledTimes(0);
+        done()
+      }
+    });
+
+  });
+
+  it('should call cdRef#markForCheck 0 times when scheduleCD is called a single time', () => {
```
Comment (Karnaukhov-kh) [inline libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts:48]:
```suggestion
  it('should call cdRef#markForCheck 1 time when scheduleCD is called a single time', () => {
```

Comment (Karnaukhov-kh) [reply libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts:48]: Test expects 1.

File: libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    const cfg = getMockNativeStrategyConfig();
+    const strategy = getStrategies(cfg)[nativeStrategyName];
+    strategy.scheduleCD();
+    expect(cfg.cdRef.markForCheck).toHaveBeenCalledTimes(1)
+    expect(cfg.cdRef.detectChanges).toHaveBeenCalledTimes(0)
+  });
+
+  it('should call cdRef#markForCheck 0 times when scheduleCD is called multiple times sync', () => {
```
Comment (Karnaukhov-kh) [inline libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts:56]:
```suggestion
  it('should call cdRef#markForCheck 4 times when scheduleCD is called multiple times sync', () => {
```

Comment (Karnaukhov-kh) [reply libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts:56]: Test expects 4

File: libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  });
+
+  it('should call cdRef#markForCheck 6 times when scheduleCD or rxScheduleCD is called', (done) => {
+    const cfg = getMockNativeStrategyConfig();
+    const strategy = getStrategies(cfg)[nativeStrategyName];
+    strategy.detectChanges = createSpy('detectChanges')
+    strategy.scheduleCD();
+    strategy.rxScheduleCD(from([0,1,2,3,4,5])).subscribe({
+      complete: () => {
+        expect(cfg.cdRef.markForCheck).toHaveBeenCalledTimes(7);
```
Comment (Karnaukhov-kh) [inline libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts:74]:
```suggestion
        expect(cfg.cdRef.markForCheck).toHaveBeenCalledTimes(6);
```

Comment (Karnaukhov-kh) [reply libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts:74]: There will be 6 emissions, right?

File: libs/template/spec/render-strategies/render-strategies/noop.strategy.spec.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
+// tslint:disable-next-line:nx-enforce-module-boundaries
+import { mockConsole } from '@test-helpers';
+import { getStrategies } from '@rx-angular/template';
+import { getMockNoopStrategyConfig } from '../../fixtures';
+import { from, of } from 'rxjs';
+import createSpy = jasmine.createSpy;
+
+const noopStrategyName = 'noop';
```
Comment (Karnaukhov-kh) [inline libs/template/spec/render-strategies/render-strategies/noop.strategy.spec.ts:8]: Same generic name thing here. `strategyName` / `testedStrategyName`?

File: libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    const strategy = getStrategies(getMockNativeStrategyConfig())[nativeStrategyName];
+    expect(strategy).toBeDefined();
+  });
+
+  it(`should have ${nativeStrategyName} as name`, () => {
+    const strategy = getStrategies(getMockNativeStrategyConfig())[nativeStrategyName];
+    expect(strategy.name).toBe(nativeStrategyName );
+  });
+
+  it('should call cdRef#markForCheck 1 time when rxScheduleCD is used with a single sync emission', (done) => {
```
Comment (Karnaukhov-kh) [inline libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts:23]: Should we separate checks for `markForCheck` & `detectChanges` into different tests? (applicable for all tests)

File: libs/template/spec/render-strategies/render-strategies/noop.strategy.spec.ts [hunk 8]
```diff
@@ -0,0 +1,67 @@
+// tslint:disable-next-line:nx-enforce-module-boundaries
+import { mockConsole } from '@test-helpers';
+import { getStrategies } from '@rx-angular/template';
+import { getMockNoopStrategyConfig } from '../../fixtures';
+import { from, of } from 'rxjs';
+import createSpy = jasmine.createSpy;
+
```
Comment (Karnaukhov-kh) [inline libs/template/spec/render-strategies/render-strategies/noop.strategy.spec.ts:7]:
```suggestion

/**
 * NOOP STRATEGY
 * Doing nothing. Should not trigger change detection
 */

```

File: libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
+// tslint:disable-next-line:nx-enforce-module-boundaries
+import { mockConsole } from '@test-helpers';
+import { getStrategies } from '@rx-angular/template';
+import { getMockNativeStrategyConfig } from '../../fixtures';
+import { from, of } from 'rxjs';
+import createSpy = jasmine.createSpy;
+
```
Comment (Karnaukhov-kh) [inline libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts:7]:
```suggestion

/**
 * NATIVE STRATEGY
 * Doing the same thing as Angular natively. Based on markForCheck.
 */
 
```

Comment (Karnaukhov-kh) [reply libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts:7]: Comments will just help to understand what change detection mechanism strategy is based on.

File: libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts (hunk 2 above)
Comment (BioPhoton) [reply libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts:8]: Will change it!

Comment (BioPhoton) [reply libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts:8]: Done!

File: libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts (hunk 7 above)
Comment (BioPhoton) [reply libs/template/spec/render-strategies/render-strategies/native.strategy.spec.ts:28]: Good idea!

File: libs/template/spec/render-strategies/render-strategies/noop.strategy.spec.ts (hunk 6 above)
Comment (BioPhoton) [reply libs/template/spec/render-strategies/render-strategies/noop.strategy.spec.ts:13]: Done

File: libs/template/spec/fixtures/mock-strategies.ts [hunk 10]
```diff
@@ -1,13 +1,7 @@
...
 import { MockChangeDetectorRef } from './fixtures';
 
-export function getMockNoopStrategyConfig() {
-  return {
-    cdRef: (new MockChangeDetectorRef() as any) as ChangeDetectorRef,
-  };
-}
-
-export function getMockNativeStrategyConfig() {
+export function getMockStrategyConfig() {
```
Comment (Karnaukhov-kh) [inline libs/template/spec/fixtures/mock-strategies.ts:4]: Nice!

File: libs/template/spec/render-strategies/render-strategies/local.strategy.spec.ts [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+import { fakeAsync, tick } from '@angular/core/testing';
+import createSpy = jasmine.createSpy;
+
+const strategyName = 'local';
+
+describe('local Strategy', () => {
+  // beforeAll(() => mockConsole());
+
+  it('should be present in strategies map', () => {
+    const strategy = getStrategies(getMockStrategyConfig())[strategyName];
```
Comment (Karnaukhov-kh) [inline libs/template/spec/render-strategies/render-strategies/local.strategy.spec.ts:15]: You just generalized `getMockStrategyConfig` :).  Can this `getStrategies(getMockStrategyConfig())` be a single fixture/helper?

File: libs/template/spec/render-strategies/render-strategies/local.strategy.spec.ts [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+      const cfg = getMockStrategyConfig();
+      const strategy = getStrategies(cfg)[strategyName];
+      strategy.scheduleCD();
+      tick(100);
+      expect(cfg.cdRef.markForCheck).toHaveBeenCalledTimes(0);
+      expect(cfg.cdRef.detectChanges).toHaveBeenCalledTimes(1);
+    })
+  );
+
+  it('should call cdRef#markForCheck 4 times when scheduleCD is called multiple times sync', (done) => {
```
Comment (Karnaukhov-kh) [inline libs/template/spec/render-strategies/render-strategies/local.strategy.spec.ts:59]:
```suggestion
  it('should call cdRef#detectChanges 4 times when scheduleCD is called multiple times sync', (done) => {
```

File: libs/template/spec/render-strategies/render-strategies/local.strategy.spec.ts [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    strategy.scheduleCD();
+    strategy.scheduleCD();
+    setTimeout(() => {
+      expect(cfg.cdRef.markForCheck).toHaveBeenCalledTimes(0);
+      expect(cfg.cdRef.detectChanges).toHaveBeenCalledTimes(4);
+      done();
+    });
+  });
+
+  it('should call cdRef#markForCheck 6 times when scheduleCD or rxScheduleCD is called', (done) => {
```
Comment (Karnaukhov-kh) [inline libs/template/spec/render-strategies/render-strategies/local.strategy.spec.ts:73]:
```suggestion
  it('should call cdRef#detectChanges 1 time when scheduleCD or rxScheduleCD is called', (done) => {
```

File: libs/template/spec/render-strategies/render-strategies/local.strategy.spec.ts (hunk 11 above)
Comment (BioPhoton) [reply libs/template/spec/render-strategies/render-strategies/local.strategy.spec.ts:19]: Unfortunately, I have to create in the test as I need a new instance to count the spy calls.

File: libs/template/spec/fixtures/emissions.ts [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+    .fill(0)
+    .forEach(
+      fakeAsync(() => {
+      fn(...args);
+      tick(100);
+    }));
+};
+
+export function singleEmissionMarkForCheck(done) {
+  return (strategyName: string, numCalls: number) => {
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply libs/template/spec/fixtures/emissions.ts]: Done

**Comment by Karnaukhov-kh** [comment]
Added refactoring suggestion. It's a single commit. Feel free to revert :).

The main goal - make stuff generic. Here `callsExpectations` object is introduced where you can set how many times you expect `markForCheck`, `detectChanges`, `detach` or `reattach` should be called. Expect statements are in `checkExpectations` function in `fixtures.ts`

File: libs/template/src/lib/render-strategies/strategies/local.strategy.ts [hunk 15]
```diff
@@ -73,6 +73,7 @@ export function createLocalStrategy<T>(
   const tick = priorityTickMap[priority];
 
   const renderMethod = () => {
+    console.log('local renderMethod');
```
Comment (kajetansw) [inline libs/template/src/lib/render-strategies/strategies/local.strategy.ts:76]: Please don't forget to remove it 😉

Comment (BioPhoton) [reply libs/template/src/lib/render-strategies/strategies/local.strategy.ts]: removed

**Pull request merged**